lib/load-config.js

```diff
@@ -22,30 +22,131 @@ const CONFIG_FILE_NAMES = Object.freeze([
   'nyc.config.cjs'
 ])
 
-function loadConfigFile (path) {
-  let config = {}
-  const jsExts = ['.js', '.cjs']
-  const ymlExts = ['.yml', '.yaml']
-  const fileName = basename(path)
+const EXTS_JS_VALIDATOR = /\.c?js$/i
+const EXTS_YAML_VALIDATOR = /\.ya?ml$/i
 
-  const ext = extname(path).toLowerCase()
-  if (jsExts.includes(ext)) {
-    config = require(path)
-  } else if (ymlExts.includes(ext)) {
-    config = require('js-yaml').load(readFileSync(path, 'utf8'))
-  } else if (ext === '.json' || fileName.slice(-2) === 'rc') {
-    config = JSON.parse(readFileSync(path, 'utf8'))
+class UnsupportedFileTypeError extends Error {
+  constructor (path) {
+    super(
+      `Unsupported file type ${extname(
+        path
+      ).toLowerCase()} while reading file ${path}`
+    )
+    this.name = UnsupportedFileTypeError.name
+  }
+}
+
+class ConfigParsingError extends Error {
+  constructor (path, errorDetails, originalError) {
+    super(
+      `Error loading configuration from file ${path}: ${errorDetails}` +
+        (typeof originalError === 'object' && originalError instanceof Error
+          ? ` Original error: ${originalError.message}`
+          : '')
+    )
+
+    this.name = UnsupportedFileTypeError.name
+
+    if (originalError instanceof Error) {
+      this.stack = originalError.stack
+    }
   }
+}
+
+const NO_EXPORTS = Symbol('no exports')
+
+function loadConfigFile (path, _di) {
+  const di = {
+    readFile: (path) => readFileSync(path, 'utf8'),
+    readJs: (path) => require(path),
+    ..._di
+  }
+
+  let config
+
+  const fileName = basename(path)
+  const ext = extname(path).toLowerCase()
+
+  if (EXTS_YAML_VALIDATOR.test(ext)) {
```

Comment:
Refactored back to the original code.  The regular expression engine is a heavy-weight process that should be
called on only when necessary.

[Addressed](https://gist.github.com/mcknasty/80c8c814864241de05987dc3cae5fd54#file-parse-args-patch-diff-L123) in review branch.

Reply:
My preference is to go without the regex.   The ext is mutated to lowercase on line 68.  All the test cases pass on the review branch atm.  Going to resolve both of these.